src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -18,56 +19,51 @@ internal sealed class SnapshotManager
             /// </summary>
             private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
             /// <summary>
-            /// The dictionary key corresponds to Syntax position. The dictionary should be sorted in descending order.
+            /// The snapshot array should be sorted in ascending order in order for the binary search algorithm to
+            /// function correctly.
             /// </summary>
-            private readonly ImmutableSortedDictionary<int, Snapshot> _incrementalSnapshots;
+            private readonly ImmutableArray<(int position, Snapshot snapshot)> _incrementalSnapshots;
 
-            private SnapshotManager(ImmutableArray<SharedWalkerState> walkerGlobalStates, ImmutableSortedDictionary<int, Snapshot> incrementalSnapshots)
+            private SnapshotManager(ImmutableArray<SharedWalkerState> walkerGlobalStates, ImmutableArray<(int position, Snapshot snapshot)> incrementalSnapshots)
             {
                 _walkerGlobalStates = walkerGlobalStates;
                 _incrementalSnapshots = incrementalSnapshots;
```

Comment:
>incrementalSnapshots [](start = 40, length = 20)

Consider adding an assert that the sequence is properly sorted and has no duplicates